File: GO/GrandePalindromoPrimo.go
/**
 * @version GO 1.19.1
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

import (
	"fmt"
	"io"
	"math"
	"net/http"
	"strconv"
	"strings"
)

/**
 * Encontre o primeiro primo palíndromo de N dígitos na expansão decimal do π (3,1415…)
 * Carrega trilhoes de dados do PI https://pi.delivery
 */
func findPrimePalindromeInPi(startPoint string, decimal int) string {
	fmt.Println("Executando teste com posicao inicial do PI em: " + startPoint + " para " + strconv.Itoa(decimal) + " digitos")
	pos, _ := strconv.Atoi(startPoint)

	resp, _ := http.Get("https://api.pi.delivery/v1/pi?start=" + startPoint + "&numberOfDigits=1000")
	body, _ := io.ReadAll(resp.Body)
	defer resp.Body.Close()
	number := string(body)

	//remove os caracteres json
	number = number[12 : len(number)-3]

	//percorrer o decimal do pi
	for i := 0; i < len(number)-decimal; i++ {
		big := number[i : i+decimal]

		if preCheck(big) {
			if isPalimdrome(big) {
				fmt.Println("Palindrome: ", big)

				if isPrime(big) {
					fmt.Println("Primo: ", big)
					fmt.Println("Tamanho: ", decimal)
					fmt.Println("Chamada de API n.: ", startPoint)
					fmt.Println("Indice da chave: ", pos+i)

					return big
				} else {
					fmt.Println("mas nao eh primo!")
				}
			}
		}
	}

	//verificar fim do arquivo de trilhoes para interromper looping.
	if strings.Contains(startPoint, "99999999999") {
		return "EOF"
	}

	//recursividade para chamada limitada da API.
	return findPrimePalindromeInPi(strconv.Itoa(pos+(1000-decimal)), decimal)
}

/**
 * Eliminate all invalid data to be palindrome and prime.
 */
func preCheck(number string) bool {
	//se o primeiro e ultimo digitos forem diferentes, ja nao eh palindromo
	if number[0] != number[len(number)-1] {
		return false
	}

	//primos nao iniciam ou terminam em par [se for palindromo o ultimo digito eh igual ao primeiro]
	b := []rune(number)
	if int(b[0])%2 == 0 {
		return false
	}

	//se for divisível por 2 ou 3 ou 5, não é primo!
	if len(number) < 16 {
		check, _ := strconv.Atoi(number)
		//check, _ := strconv.ParseInt(number, 10, 64)

		if check%2 == 0 || check%3 == 0 || check%5 == 0 {
			return false
		}
	}

	//se a soma dos algorismos forem divisiveis por 3 nao eh primo
	soma := 0
	for i := 0; i < len(number); i++ {
		digito, _ := strconv.Atoi(string(number[i]))

		soma += digito
	}
	return soma%3 != 0

	//Todo número palíndromo com uma quantidade par de algarismo é divisível por 11

	//nenhum erro encontrado
	//return true
}

/**
 * Check if the number is palindrome.
 */
func isPalimdrome(number string) bool {
	//inverte a string para comparar com original
	reverse := []rune(number)
	for i, j := 0, len(reverse)-1; i < j; i, j = i+1, j-1 {
		reverse[i], reverse[j] = reverse[j], reverse[i]
	}

	return string(reverse) == number
}

/**
 * Check if the number is prime.
 * Adaptado de: https://www.mentebinaria.com.br/forums/topic/795-uma-otimiza%C3%A7%C3%A3o-interessante-para-testar-se-um-valor-%C3%A9-primo/
 */
func isPrime(n string) bool {
	if len(n) > 15 {
		return isOnlinePrime(n)
	}

	number, _ := strconv.ParseInt(n, 10, 64)

	//so precisamos testar até a raiz quadrada
	var raiz int64 = int64(math.Sqrt(float64(number)))
	var i int64

	for i = 3; i <= raiz; i += 2 {
		if int64(number)%i == 0 {
			return false
		}
	}

	return true
}

/**
 * solucao temporaria para verificar se palidromos acima de 15 digitos sao primos
 * https://www.numberempire.com/primenumbers.php
 */
func isOnlinePrime(number string) bool {
	resp, _ := http.Get("https://www.numberempire.com/primenumbers.php?number=" + number)
	body, _ := io.ReadAll(resp.Body)
	defer resp.Body.Close()
	html := string(body)

	//verifica o retorno
	check := strings.Index(html, number+" is a prime")

	return check != -1
}

File: JAVA/GrandePalindromoPrimo.java
/**
 * @version JAVA
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package JAVA;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class GrandePalindromoPrimo {

    public static void main(String[] args) {
        System.out.printf("Palindrome: %d\n", findPrimePalindromeInPi("128919", 9));
    }

    /**
     * Encontre o primeiro primo palíndromo de N dígitos na expansão decimal do π (3,1415…)
     * Carrega trilhoes de dados do PI https://pi.delivery
     */
    public static String findPrimePalindromeInPi(String startPoint, int decimal) {
        System.out.println("Executando teste com posicao inicial do PI em: " + startPoint + " para " + decimal + " digitos");

        try {
            URL url = new URL("https://api.pi.delivery/v1/pi?start="+startPoint+"&numberOfDigits=1000");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");

            BufferedReader in = new BufferedReader(new InputStreamReader( conn.getInputStream() ));
            String number = in.readLine();
            in.close();

            //remove os caracteres json
            number = number.substring(12, number.length()-2);

            //percorrer o decimal do pi
            for ( int i=0; i < number.length()-decimal; i++ ) {
                String big = number.substring( i, i+decimal );

                if ( preCheck( big ) ) {
                    if ( isPalimdrome( big ) ) {
                        System.out.println("Palindrome: "+big);

                        if ( isPrime( big ) ) {
                            System.out.println("Primo: "+big);
                            System.out.println("Tamanho: "+decimal);
                            System.out.println("Chamada de API n.: "+startPoint);
                            System.out.println("Indice da chave: "+(Long.parseLong(startPoint)+i));

                            return big;
                        } else {
                            System.out.println("mas nao eh primo!");
                        }
                    }
                }
            } 
        } catch ( Exception err ) {
            System.out.println( err );
            return "erro";
        }

        //verificar fim do arquivo de trilhoes para interromper looping.
        if ( startPoint.contains("99999999999") ) {
            return "EOF";
        }

        //recursividade para chamada limitada da API.
        return findPrimePalindromeInPi( String.valueOf(Long.parseLong(startPoint)+(1000-decimal)), decimal );
    }

    /**
     * Eliminate all invalid data to be palindrome and prime.
     */
    private static boolean preCheck(String number) {
        //se o primeiro e ultimo digitos forem diferentes, ja nao eh palindromo
        if ( number.charAt(0) != number.charAt(number.length()-1) ) return false;

        //primos nao iniciam ou terminam em par [se for palindromo o ultimo digito eh igual ao primeiro]
        if ( number.charAt(0)%2 == 0 ) return false;

        /** nao consegui fazer a divisao para numeros gigantes bitinteger (?) */
        if ( number.length() < 16 ) {
            //se for divisível por 2 ou 3 ou 5, não é primo!
            long check = Long.parseLong( number );

            if ( check % 2 == 0 || check % 3 == 0 || check % 5 == 0 ) return false;
        }

        //se a soma dos algorismos forem divisiveis por 3 nao eh primo
        int soma = 0;
        for ( int i = 0; i < number.length(); i++ ) {
            soma += number.charAt(i);
        }
        if ( soma % 3 == 0 ) return false;

        //nenhum erro encontrado
        return true;
    }

    /**
     * Check if the number is palindrome.
     */
    private static boolean isPalimdrome(String number) {
        StringBuilder reverse = new StringBuilder();
        reverse.append( number );
        reverse.reverse();

        return number.contentEquals( reverse );
    }

    /**
     * Check if the number is prime.
     * Adaptado de: https://www.mentebinaria.com.br/forums/topic/795-uma-otimiza%C3%A7%C3%A3o-interessante-para-testar-se-um-valor-%C3%A9-primo/
     */
    private static boolean isPrime(String number) throws Exception {
        //problemas com numeros muito grandes
        if ( number.length() > 15 ) {
            return isBigPrime( number );
        }

        //so precisamos testar até a raiz quadrada
        int raiz = (int) Math.sqrt(Integer.parseInt(number));
    
        for ( int i = 3; i <= raiz; i += 2 ) {
            if ( ( Integer.parseInt(number) % i ) == 0 ) return false;
        }
    
        return true;
    }

    /**
     * solucao temporaria para verificar se palidromos acima de 15 digitos sao primos
     * https://www.numberempire.com/primenumbers.php
     */
    private static boolean isBigPrime(String number) throws Exception {
        URL u = new URL("https://www.numberempire.com/primenumbers.php?number="+number);
        HttpURLConnection c = (HttpURLConnection) u.openConnection();
        c.setRequestMethod("GET");

        BufferedReader input = new BufferedReader(new InputStreamReader( c.getInputStream() ));
        String html = input.readLine();
        input.close();
        
        //verifica o retorno
        return html.contains(number+" is a prime");

        //BigInteger number = new BigInteger(n);
        //long raiz = number.sqrt().longValue();
        //for ( int i = 3; i <= raiz; i += 2 ) {
            //if ( BigInteger.valueOf(i).mod(number) == BigInteger.valueOf(0) ) return false;
        //}
        //return true;
    }
}
